Make i18n async #2

fredlahde opened the pull request:
This PR adds a possible solution for #2 . 

This solution adds an understanding of "strict" revolvers. A strict resolver resolves on the full value of `navigator.language` e.g. `en-US`. A non-strict one would resolve with `en` in this case.

If `getInnerHTML()` is called, it checks if revolvers are available and then calls `resolveI18n`. This will asynchously check the strict resolvers for any matches. If it finds a match, it will call `setInnerHtml`. If it can't find a strict match, it'll try the non strict resolvers. If they can't find a result, the function rejects.

fredlahde committed:
Make i18n async #2

- VNode: getInnerHtml() does no longer resolve i18n
- VNode: add resolvei18n(). This method asynchronously tries to resolve
i18n in the same manner getInnerHtml() did. If it finds a match, it
stores the 'raw' innerHTML (with the unresolved i18n keys) in
'rawInnerHtml' and registers an on DOM Listener, which calls
setInnerHtml with the now resolved innerHtml. I'm storing the origina
innerHTML in a field, because this way we can offer a way to re-resolve
i18n, if the locale changes. resolvei18n(), called a second time, will
use rawInnerHtml as base for resolving.
- VNode: call resolvei18n() in constructor, redirect on error
- Resolver: Make get() return a Promise
- Router: add hasRouteRegistered()

larsgrah commented on src/vdom/VNode.ts:
```diff
@@ -56,6 +58,14 @@ export class VNode implements Comparable<VNode> {
             this.attrs.push(new Attribute(kloudAppId, this.id));
         }
+
+        this.resolvei18n().catch(e => {
```
Don't route to error here, rather just let it fail. If another resolver is able to complete the request it should be fine

fredlahde replied:
Done

fredlahde committed:
Merge branch 'master' into async-i18n

larsgrah reviewed (changes requested):
There needs to be a centralized way to change the i18n after the fact. Also we do not want to trigger re-renders in the constructor of the vNode, as this would add considerable overhead and make methods like app.k behave dirty

fredlahde committed:
Make i18n resolving lazy

- Do not resolve on construct of a node, but rather on getInnerHtml()
- fuzzy matching of locales
- call setInnerHtml() directly
- ony resovle if locale has changed since the last time we resolved

fredlahde committed:
Package lock

larsgrah commented on src/vdom/VNode.ts:
```diff
@@ -141,25 +144,51 @@ export class VNode implements Comparable<VNode> {
+                })
+                .map(it => reflect(it))
+
+            Promise.all(resolver)
```
We want to use the first non-rejected value here. Consider using a "race to success" function where the first successful promise is greedily returned

fredlahde committed:
Change the flow in which i18n is resolved

- First check for strict resolvers
- If none of them resovles, check non strict ones
- If they also not resolve, reject

larsgrah commented on src/vdom/VNode.ts:
```diff
@@ -1,4 +1,4 @@
-import { Comparable, arraysEquals, Stringparser, dataRegex } from './Common';
+import { Comparable, arraysEquals, Stringparser, dataRegex, reflect, raceToSuccess } from './Common';
```
remove the reflect import here?

larsgrah commented on src/vdom/Common.ts:
```diff
@@ -122,3 +122,31 @@ const parse = (node: Element, app: VApp): VNode => {
+ * Takes a promise and returns one, wich always resovles, regardless of the outcome of the given promise
+ * @param promise
+ */
+export function reflect<T>(promise: Promise<T>): Promise<T> {
```
Do we need this with raceToSucess being present?

larsgrah commented on src/vdom/VNode.ts:
```diff
@@ -141,25 +144,61 @@ export class VNode implements Comparable<VNode> {
+                });
+
+            raceToSuccess(strictResolver)
+                .then(match => {
```
We have duplicate code here, consider making this into a function

fredlahde committed:
Remove reflect(), refactor duplicate code in resovleI18n

larsgrah reviewed (approved):
Looks good
